src/ripple/core/impl/SociDB.cpp

```diff
@@ -196,44 +197,33 @@ namespace {
     is the default behavior of sqlite. We may be able to remove this
     class.
 */
+
 class WALCheckpointer : public Checkpointer
 {
 public:
-    WALCheckpointer(sqlite_api::sqlite3& conn, JobQueue& q, Logs& logs)
-        : conn_(conn), jobQueue_(q), j_(logs.journal("WALCheckpointer"))
+    WALCheckpointer(
+        std::uint64_t id,
+        sqlite_api::sqlite3& conn,
+        JobQueue& q,
+        Logs& logs)
+        : Id_(id)
+        , conn_(conn)
+        , jobQueue_(q)
+        , j_(logs.journal("WALCheckpointer"))
     {
-        sqlite_api::sqlite3_wal_hook(&conn_, &sqliteWALHook, this);
+        sqlite_api::sqlite3_wal_hook(&conn_, &sqliteWALHook, (void*)Id_);
```

Comment:
I'd prefer if we used: `reinterpret_cast<void*>(id_)` instead because this cast is "sufficiently weird" that it should stand out.

Reply:
Also, do we **enforce** the SQLite requirements for [`sqlite3_wal_hook`](https://www.sqlite.org/c3ref/wal_hook.html)?

>**A single database handle may have at most a single write-ahead log callback registered at one time. Calling sqlite3_wal_hook() replaces any previously registered write-ahead log callback.** Note that the sqlite3_wal_autocheckpoint() interface and the wal_autocheckpoint pragma both invoke sqlite3_wal_hook() and will overwrite any prior sqlite3_wal_hook() settings.

Reply:
Fixed the cast in `1a499c7b4 Styles fixes and use correct int type to cast to pointer`

I don't think it's an error to call `DatabaseCon::setupCheckpointing` more than once, it would just replace the current checkpointer and everything will continue to work. However, it doesn't really make sense to ever do that. It shouldn't be too hard to make `DatabaseCon::setupCheckpointing` private and make it part of the constructor. I'll look at doing that.

Reply:
@nbougalis I pushed a patch that ensures `setupCheckpointing` is called at most once: `3a6e5ab8a Ensure checkpointing is setup at most once` Take a look. It doubles the number of constructors, but with delegating constructors that's not too bad.